### PR TITLE
Ray differentials fixes and improvements.

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp [hunk 1]
```diff
@@ -579,6 +579,29 @@ void ShadingPoint::compute_world_space_partial_derivatives() const
     }
 }
 
+namespace
+{
+
+template <typename T>
```
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/renderer/kernel/shading/shadingpoint.cpp:585]: Also `intersect_ray_plane()` would be a little more consistent.

File: src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp [hunk 2]
```diff
@@ -40,13 +40,48 @@ using namespace foundation;
 
 namespace renderer
 {
+namespace
+{
+
+void compute_normal_derivatives(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp:46]: Indent content of anonymous namespaces.

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (est77) [reply src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: done

File: src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp (hunk 2 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp]: done

File: src/appleseed/renderer/modeling/bsdf/diffusebtdf.cpp [hunk 3]
```diff
@@ -145,6 +145,7 @@ namespace
                 sample.m_value.m_beauty = sample.m_value.m_diffuse;
 
                 sample.m_min_roughness = 1.0f;
+                sample.compute_diffuse_differentials(outgoing);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/diffusebtdf.cpp:148]: Given that we're dealing with transmission here, are the differentials correct? Just wondering about the lack of reflected vs transmitted versions of the diffuse case.

Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/diffusebtdf.cpp:148]: Diffuse differentials are always 1/25 of the sphere in the direction of the outgoing vector.
There is no special case for transmitted diffuse differentials.

**Pull request merged**